### DISCUSSION TITLE
Gitlab CodeQuality Reports

### DISCUSSION THREAD

**Discussion by jhannink** [detekt/detekt#3350] [Q&A] [answered]
Is there already a plugin generating reports that could be used to upload code quality metrics for a gitlab CI / CD job?
The json spec for this is a subset of the codeclimate report spec and described here: https://docs.gitlab.com/ee/user/project/merge_requests/code_quality.html#implementing-a-custom-tool

Thanks :)

**Comment by picklebento** [comment, marked as answer]
Thanks for using the Q&A discussion! We are adding SARIF support in #3045. It looks like Gitlab is also exploring support SARIF format https://gitlab.com/gitlab-org/gitlab/-/issues/118496.

If you would like to customize the report, you may want to check out the example [here](https://github.com/detekt/detekt/blob/master/detekt-sample-extensions/src/main/kotlin/io/gitlab/arturbosch/detekt/sample/extensions/reports/QualifiedNamesOutputReport.kt)
Contributions are also welcome if you would like to standardize it into the `detekt-report-codeclimate` module.

**Reply by cromefire** [reply]
I've updated the GitLab plugin meanwhile to also support basically everything in the code climate spec, so if you're interested, I could contribute it as code climate report. With that done, if I understand it correctly, you'd "only" need to create a docker container with [some specific](https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#engine-execution) properties, implement their `include_paths`, maybe add a keys to the config for cli options and that should be sufficient.

I think the plugin for GitLab still makes sense, as that analyzer probably couldn't use type information and the container would need to be integrated into code climate first (as GitLab's format is slightly different and it uses "normal" code climate analyzer via so converter it seems), but it might be a huge benefit for users who use code climate which doesn't work on Kotlin currently as far as I'm aware.

**Comment by MaksimMyshkin** [comment]
There is plugin that generates GitLab formatted reports https://gitlab.com/cromefire_/detekt-gitlab-report . Working nice on our project.

**Reply by cortinico** [reply]
Do you wish to add it to the Detekt Readme?

**Reply by MaksimMyshkin** [reply]
It is already there. I think I just found out about this from the Readme) Left a comment in case someone bumps into this comment while searching like me. It also looks like it is a fairly new tool. When I first tried to integrate detekt with GitLab, I couldn't find anything.
![image](https://user-images.githubusercontent.com/16517593/127467616-a416bbb5-59c5-48f9-b0a3-4e33d915cc4c.png)

**Reply by cromefire** [reply]
PSA: I've now published a stable release, please consider upgrading to `0.1.0` (which is basically the same, but not as a SNAPSHOT) or `0.2.1` if you use the latest version of detekt. Those releases are stable and never change (opposed to snapshots, which automatically updates on any commit to master).
(sorry for the direct comment, but just wanna make sure nothing breaks as my initial release practice wasn't particularly good practice...)

**Reply by zilvinas-t** [reply]
Hello. The `gitlab.json` file mentioned in the documentation is not generated. I have a case with multimodule project and I want to generate a one report for all modules. Maybe there is a step I'm missing? The code:

detekt.gradle:

```
apply plugin: "io.gitlab.arturbosch.detekt"

detekt {
    buildUponDefaultConfig = true
    ignoreFailures = false
    allRules = false
    parallel = true
    basePath = projectDir
    config = files("$rootDir/config/detekt/detekt.yml")
    baseline = file("$rootDir/config/detekt/baseline.xml")
}

dependencies {
    detektPlugins("com.gitlab.cromefire:detekt-gitlab-report:0.2.1")
}
```

build.gradle (project level):

```
def projectSource = file(projectDir)
def configFile = files("$rootDir/config/detekt/detekt.yml")
def baselineFile = file("$rootDir/config/detekt/baseline.xml")
def kotlinFiles = "**/*.kt"
def resourceFiles = "**/resources/**"
def buildFiles = "**/build/**"

tasks.register("detektAll", io.gitlab.arturbosch.detekt.Detekt) {
    def autoFix = project.hasProperty('detektAutoFix')

    description = "Custom DETEKT build for all modules"
    autoCorrect = autoFix
    setSource(projectSource)
    baseline.set(baselineFile)
    config.setFrom(configFile)
    include(kotlinFiles)
    exclude(resourceFiles, buildFiles)
    reports {
        custom {
            reportId = "DetektGitlabReport"
            destination = file("$buildDir/reports/detekt/gitlab.json")
        }
        html.enabled = true
        xml.enabled = true
        txt.enabled = false
        sarif.enabled = false
    }
}
```

**Reply by cortinico** [reply]
@zilvinas-t You should open an issue on Gitlab's report repo: https://gitlab.com/cromefire_/detekt-gitlab-report

**Reply by cromefire** [reply]
I'm happy to sort it out and get it fixed if necessary.

**Reply by MaksimMyshkin** [reply]
In our project we just merge reports from all modules with jq: 
```yaml
after_script:
    - apk add jq
    - find . -path "*/target/detekt-reports/gitlab.json" | xargs jq -s '[.[][]]' -c > detekt-report.json`
```
